### ISSUE TITLE
Retrain CNNs

### ISSUE THREAD

**Issue by AliceDinh** [faceterteam/PayCards_Android#8]
I want to retrain CNNs using my samples then test by my own inputs but I don't know how, please show me.
Thanks

**Comment by tchernitski** [comment]
I am afraid I can't show how to retrain CNN in single comment. It's quite broad question. There are a lot of tutorials in internet. 

Just to let you know. We used deep learning framework http://caffe.berkeleyvision.org/ but you are free to replace it with other deep learning framework. There are two types of networks in the project are used. The first one is for objects localization task and second is for classification task.

**Comment by AliceDinh** [comment]
Hi,
Sorry if I disturb you much, I am just a beginner in this field. So can you please explain to me the meaning of your models:
DateLocalizationL0.caffemodel
DateLocalizationL1.caffemodel
DateRecognition.caffemodel
NameLocalizationX.caffemodel
NameSpaceCharRecognition.caffemodel
loc_x.caffemodel
loc_y.caffemodel
NumberRecognition.caffemodel

AND

What dataset you used to train those models?
Thanks

**Comment by tchernitski** [comment]
Hi Alice,

All these files contain neural networks weights that we got as result of cnn training. It's protobuf binary format used in caffe framework.
We used our own dataset of embossed digits/letters that we collected form real bank card samples and collected in internet.

**Comment by AliceDinh** [comment]
Hi Tchernitski, thanks for your help. And is that possible to share your dataset publicly? 
Thanks

**Comment by tchernitski** [comment]
I am afraid we can't share it. The dataset contains sensitive data like card numbers and cardholder names. I would recommend you to google sites like https://colnect.com/en/bank_cards. There are a lot of images of bank card that can be used as raw data for training dataset.

**Comment by AliceDinh** [comment]
Hi Tchernitski,
Thanks for your share, I haven't seen that site before, that's great help from you.
Because you cannot share your dataset, so I want to ask something else which is: 
- For each digit/letter, how many samples of them you need to train?
- Samples for each digit/letter are different ones or just one image then you convert them (add noise, ...) to different ones?
- For non-embossed digits/letters, I have to build model then train from my own samples, but should I put them into the same model of embossed digits/letters. I mean should I merge the process of training of embossed digits/letters and non-embossed digits/letters into ONE model or I have to build and train into TWO separate models?

Thanks and regards,
Alice

**Comment by tchernitski** [comment]
> For each digit/letter, how many samples of them you need for training?

I don't remember exact number of samples but there were about 1500 cards, so it were about 1500*16 digit samples. Actually there were fewer samples because of a lot of digits were masked on bank cards downloaded from internet. Also take in to account that there was quite large dispersion of samples in classes. Like a lot of "4" and "5" and "0" digits, but fewer of others. 

> Samples for each digit/letter are different ones or just one image then you convert them (add noise, ...) to create different ones?

Definitely use data augmentation: samples shift, blur, noise etc

> For non-embossed digits/letters, I have to build model then train from my own samples, but should I put them into the same model of embossed digits/letters. I mean should I merge the process of training of embossed digits/letters and non-embossed digits/letters into ONE model or I have to build and train into TWO separate models?

I believe you can "merge" embossed and non-embossed samples and train single model and it should work. But also you can try another approach. Train two separate embossed/non-embossed models and third one binary cnn to classify if a card is embossed or non-embossed. I am not sure which one approach is better. It needs to be tested.

> Can you provide your solver prototxt files please?

I am sorry I don't have time to retrieve training stuff of this project right now. But I remember it was simple SGD solver. Active phase of the project was 2015. There are a lot of cnn architectures appeared since that days.

**Comment by AliceDinh** [comment]
Thanks Tchernitski,

> There are a lot of cnn architectures appeared since that days.

1. which cnn you used?
2. And what is following numbers' mean? I am building model for 19-number-cards from my country, how these numbers affect and in case of 19 digits instead of 16, what other numbers should be?
```
input_dim: 16
input_dim: 1
input_dim: 41
input_dim: 29
```
Regards,
Alice

**Comment by tchernitski** [comment]
Alice, 
I would strongly recommend you to read some artice(s) about cnn and cnn training.
There are a lot of posts about it. For example http://adilmoujahid.com/posts/2016/06/introduction-deep-learning-python-caffe/
I believe it's going to be more efficient. 

We used our own architecture (it's quite old for now) that was balanced between performance and speed due to limited computation resources on mobile phones. 

input_dim: 16  // we send 16 digits at once (batch size)
input_dim: 1 // grayscale
input_dim: 41 // sample height
input_dim: 29 // sample width

num_output: 10 // final classifier layer, the output is 10 because we have only 10 digits (0,1,2 ... 9), so the output is 10 probabilities

**Comment by AliceDinh** [comment]
Thanks Tchernitski very much for your great help.

**Comment by AliceDinh** [comment]
@tchernitski 
Hi, long time haven't been contacted you.
I followed your suggestions, using SSD mobilenet to get better performance instead of using your pre-trained Caffe model. Briefly, I have done following works:
1. I collected 1000 card samples (16 digits, 19 digits)
2. I used TensorFlow Object Detection API (https://github.com/tensorflow/models/tree/master/research/object_detection) to train, test model, then export it to frozen model that will be run on mobiles (the frozen model has .pb extension)
3. The neural net I tested are Mask R-CNN, Faster R-CNN, SSD but I found only SSD can be suitable for mobiles. (SSD config file is at: https://github.com/tensorflow/models/blob/master/research/object_detection/samples/configs/ssd_mobilenet_v1_pets.config)
4. Result: the SSD model can detect some digits on card (<16 digits, or <19 digits), whatever objects are detected, they are labeled correctly. So I am not worry about the classification phase, I am worry about the localization phase.
Here is an example
![43444442-ecd9b820-94cd-11e8-93e6-e4920182e398](https://user-images.githubusercontent.com/20389638/43751687-219c2e62-9a28-11e8-9b7f-f86d7f903e7c.png)

So, I have some questions that really need your help:
1. Please have look at the SSD configuration file and advise me which parameters may affect the accuracy. You may not work with this at the moment but with the experience you had, I believe you will give me very useful advise.
2. The input images, I mean the images for dataset before training, does the input-size affect the training? SSD resizing step will resize inputs into 300x300, should I take 300x300 inputs, or whatever sizes?
Thanks @tchernitski  very much and I appreciate your help.
